lib/IDE/Refactoring.cpp

```diff
@@ -2602,6 +2602,109 @@ bool RefactoringActionLocalizeString::performChange() {
+                            SourceLoc targetLocation) {
+  
+  if (memberTypeVector.size() > 0) {
+    EditConsumer.accept(SM, targetLocation, "\ninternal init(");
+    
+    for (size_t i = 0; i < memberTypeVector.size(); i++) {
+      EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " +
+                          memberTypeVector[i]);
+      
+      if (i != memberTypeVector.size() - 1) {
+        EditConsumer.accept(SM, targetLocation, ", ");
+      }
+    }
+    
+    EditConsumer.accept(SM, targetLocation, ") {\n\n");
+    
+    for (auto varName: memberNameVector) {
+      EditConsumer.accept(SM, targetLocation,
+                          "self." + varName + " = " + varName + "\n");
+    }
+    
+    EditConsumer.accept(SM, targetLocation, "\n}\n");
+  }
+}
+  
+static ClassDecl* collectMembersForInit(ResolvedCursorInfo CursorInfo,
+                           SmallVectorImpl<std::string>& memberNameVector,
+                           SmallVectorImpl<std::string>& memberTypeVector) {
+  
+  if (CursorInfo.Kind != CursorInfoKind::ValueRef && !CursorInfo.ValueD)
+    return nullptr;
+  
+  ClassDecl *classDecl = dyn_cast<ClassDecl>(CursorInfo.ValueD);
+  if (!classDecl || classDecl->getStoredProperties().empty() ||
+      CursorInfo.IsRef) {
+    return nullptr;
+  }
+  
+  for (auto varDecl : classDecl->getStoredProperties()) {
+    auto parentPatternBinding = varDecl->getParentPatternBinding();
+    if (!parentPatternBinding)
+      continue;
+    
+    auto varDeclIndex =
+      parentPatternBinding->getPatternEntryIndexForVarDecl(varDecl);
+    
+    if (auto init = varDecl->getParentPatternBinding()->getInit(varDeclIndex)) {
+      if (init->getStartLoc().isValid())
+        continue;
+    }
+    
+    StringRef memberName = varDecl->getName().str();
+    memberNameVector.push_back(memberName.str());
+    
+    std::string memberType = varDecl->getType().getString();
+    memberTypeVector.push_back(memberType);
+  }
+  
+  return classDecl;
+}
+
+bool RefactoringActionMemberwiseInitLocalRefactoring::
+isApplicable(ResolvedCursorInfo Tok, DiagnosticEngine &Diag) {
+  
+  SmallVector<std::string, 5> memberNameVector;
+  SmallVector<std::string, 5> memberTypeVector;
+  
+  collectMembersForInit(Tok, memberNameVector, memberTypeVector);
+  
+  return !memberTypeVector.empty();
+}
+    
+bool RefactoringActionMemberwiseInitLocalRefactoring::performChange() {
+  
+  if (!CursorInfo.ValueD)
+    return true;
+  
+  SmallVector<std::string, 5> memberNameVector;
+  SmallVector<std::string, 5> memberTypeVector;
+  
+  auto classDecl = collectMembersForInit(CursorInfo, memberNameVector,
+                                         memberTypeVector);
+  if (!classDecl)
+    return true;
+  
+  swift::SourceLoc bracesStart = classDecl->getBraces().Start;
+  if (!bracesStart.isValid())
+    return true;
+  
+  swift::SourceLoc targetLocation = bracesStart.getAdvancedLoc(1);
+  if (!targetLocation.isValid())
+    return true;
```

Comment:
Move these two source loc checks into `collectMembersForInit` and return `targetLocation` instead of `classDecl` from `collectMembersForInit`. The behavior should be a non-empty `memberNameVector` implies valid location for inserting the initializer and we're sure to perform change in users' code.